File: package/android/src/main/java/com/margelo/nitro/realfetch/HybridDuplexStream.kt
package com.margelo.nitro.realfetch

import java.io.PipedInputStream
import java.io.PipedOutputStream

class HybridDuplexStream : HybridDuplexStreamSpec() {
  companion object {
    private const val BUFFER_SIZE = 64 * 1024
  }

  private val pipeIn = PipedInputStream(BUFFER_SIZE)
  private val pipeOut = PipedOutputStream(pipeIn)

  override var inputStream: HybridInputStreamSpec = HybridInputStream(pipeIn)
  override var outputStream: HybridOutputStreamSpec = HybridOutputStream(pipeOut)

  override val memorySize: Long = 0L

  fun close() {
    try {
      pipeOut.close()
      pipeIn.close()
    }
    catch (e: Exception) {
      System.err.println("Error closing duplex stream: ${e.message}")
    }
  }
}

File: package/ios/HybridDuplexStream.swift
import Foundation

class HybridDuplexStream: HybridDuplexStreamSpec {
  private static let bufferSize = 64 * 1024

  var inputStream: (any HybridInputStreamSpec)
  var outputStream: (any HybridOutputStreamSpec)

  override init() {
    var inputStreamRef: InputStream? = InputStream()
    var outputStreamRef: OutputStream? = OutputStream(toMemory: ())

    Stream.getBoundStreams(
      withBufferSize: HybridDuplexStream.bufferSize,
      inputStream: &inputStreamRef,
      outputStream: &outputStreamRef)

    guard let inputStreamRef, let outputStreamRef else {
      fatalError("Could not create streams")
    }

    inputStream = HybridInputStream(stream: inputStreamRef)
    outputStream = HybridOutputStream(stream: outputStreamRef)
  }

  var memorySize: Int {
    return 0
  }
}
